### PR TITLE
#878 -- Refactor DeviceProfile endpoints

### PR DESCRIPTION
**Pull request by jdharms**
#878 Refactors DeviceProfile endpoints and models to abstract
away the MongoDB specific details.  Also refactors DeviceProfile
database-related functions to return values instead of mutating
pointer arguments.

Signed-off-by: Daniel Harms <jdharms@gmail.com>

### CODE REVIEW

File: internal/pkg/db/mongo/metadata.go [hunk 1]
```diff
@@ -515,71 +558,88 @@ func (m MongoClient) GetDeviceProfile(d *contract.DeviceProfile, q bson.M) error
...
 
-func (m MongoClient) UpdateDeviceProfile(dp *contract.DeviceProfile) error {
+func (m MongoClient) UpdateDeviceProfile(dp contract.DeviceProfile) error {
 	s := m.session.Copy()
 	defer s.Close()
 	c := s.DB(m.database.Name).C(db.DeviceProfile)
 
-	mdp := mongoDeviceProfile{DeviceProfile: *dp}
-	mdp.Modified = db.MakeTimestamp()
+	dp.Modified = db.MakeTimestamp()
```
[the remark this answers is not in the document]
Comment (michaelestrin) [reply internal/pkg/db/mongo/metadata.go]: Actually, this is fine as is.

File: internal/pkg/db/mongo/metadata.go [hunk 2]
```diff
@@ -515,71 +558,88 @@ func (m MongoClient) GetDeviceProfile(d *contract.DeviceProfile, q bson.M) error
...
 	}
 	for i := 0; i < len(dp.Commands); i++ {
 		if newId, err := m.AddCommand(dp.Commands[i]); err != nil {
-			return err
+			return "", err
 		} else {
 			dp.Commands[i].Id = newId
 		}
 	}
 	ts := db.MakeTimestamp()
```
[the remark this answers is not in the document]
Comment (jdharms) [reply internal/pkg/db/mongo/metadata.go]: Not quite sure how that would work

[the remark this answers is not in the document]
Comment (michaelestrin) [reply internal/pkg/db/mongo/metadata.go]: If memory serves, my reference was to the timestamp assignments.  And I vaguely recall adding a subsequent comment to the effect that my original comment was wrong.  I also recall marking those comments are resolved.  So nothing to see here... please move along. :)

**Comment by jdharms** [comment]
> In `src/github.com/edgexfoundry/edgex-go/pkg/models/`, remove bson tags from `profileproperty.go`, `profileresource.go`, `propertyvalue.go`, `resourceoperation.go`, and `units.go`.

Missed this one--just updated again.

File: internal/pkg/db/mongo/metadata.go [hunk 3]
```diff
@@ -455,131 +455,156 @@ func (m MongoClient) GetDevice(d *contract.Device, q bson.M) error {
...
+	err = col.Find(q).Sort("queryts").All(&dps)
 	if err != nil {
-		return err
+		return []contract.DeviceProfile{}, err
 	}
 
-	*d = []contract.DeviceProfile{}
-	for _, mdp := range mdps {
-		*d = append(*d, mdp.DeviceProfile)
+	for _, dp := range dps {
```
Comment (michaelestrin) [inline internal/pkg/db/mongo/metadata.go]: This bit me earlier this week.  You need to initialize `cdps` to an empty slice here.  It defaults to nil.  Which will be returned if `dps` is empty.

**Comment by jdharms** [comment]
These are failures present on master as well--all have to do with a file parameter to the deviceprofile/uploadfile route.  I think this is a problem with newman or our configuration--not sure which.

**Comment by tsconn23** [comment]
@jdharms The only one in that list I'd be concerned about is 157.  That one doesn't use a file and instead POSTs a JSON string which should be workable. The rest of them are all those tests dealing with file uploads that we've discussed with the IOTech folks in the past. I wouldn't let those block this PR. Let us know what you find on 157.

**Comment by jdharms** [comment]
@tsconn23 I believe @michaelestrin was using stale tests.  Updates in: https://github.com/edgexfoundry/blackbox-testing/pull/135 should make the test green again.

**Pull request merged**